corehq/util/pagination.py

```diff
@@ -71,8 +84,11 @@ def paginate_function(get_data_fn, next_args_fn, event_handler=PaginationEventHa
         total_emitted += len_results
         event_handler.page_end(total_emitted, datetime.utcnow() - start_time, *args, **kwargs)
 
-        if len_results:
-            args, kwargs = next_args_fn(item, *args, **kwargs)
+        item = item if len_results else None
+        try:
+            args, kwargs = args_provider.get_next_args(item, *args, **kwargs)
```

Comment:
also changed so that iteration continues until `get_next_args` raises `StopIteration`. The purpose of this is to support use cases like `DocTypeIterator` which iterates over multiple doc types in succession (basically a nested loop).